### ISSUE TITLE
ACL match ip pattern does not recognise IPv6 addresses in square brackets

### ISSUE THREAD

**Issue by Puliyo** [haproxy/haproxy#2133]
### Detailed Description of the Problem

IPv6 can come in square bracket form.
`[2001:db8:85a3:8d3:1319:8a2e:370:7348]:443`

ACL match IP pattern does not pick this up as valid IPv6 format.

Related issue: https://github.com/haproxy/haproxy/issues/2054

### Expected Behavior

Square bracket IPv6 with/without port to be captured by ACL match IP pattern.

### Steps to Reproduce the Behavior

1. Create ACL to capture all IPv6: `acl isipv6 req.hdr(host) -m ip ::/0`
2. Create simple rule to deny if not ipv6: `http-request deny deny_status 429 hdr Denial-Reason "not ipv6" if !isipv6`
3. Send request to haproxy with Host header containing ipv6 with square bracket

### What is your configuration?

```haproxy
defaults
    mode http
```
```


### Output of `haproxy -vv`

```plain
HAProxy version 2.7.6-1ppa1~jammy 2023/04/01 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2024.
Known bugs: http://www.haproxy.org/bugs/bugs-2.7.6.html
Running on: Linux 5.15.0-1027-oracle #33-Ubuntu SMP Fri Jan 6 16:30:16 UTC 2023 aarch64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_OT=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL +OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC +RT +SHM_OPEN +SLZ -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=2).
Built with OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
Running on OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with OpenTracing support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.39 2021-10-29
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [  OT] opentracing
        [SPOE] spoe
        [TRACE] trace
```
```


### Last Outputs and Backtraces

_No response_

### Additional Information

_No response_

**Comment by lukastribus** [comment]
> IPv6 can come in square bracket form.
> `[2001:db8:85a3:8d3:1319:8a2e:370:7348]:443`

I don't think we will ever match IP address with trailing port, I don't think `192.0.2.0:80` matches `0.0.0.0/0`.

But without trailing port, yes, `-m ip` should match this.

**Comment by Puliyo** [comment]
Current haproxy matches with port.

```
acl blockip4 req.hdr(host) -m ip 0.0.0.0/0
http-request deny deny_status 429 hdr Denial-Reason "ipv4!" if blockip4
```

```
$ curl -v http://127.0.0.1:8080 -H 'host: 127.0.0.1:12345'
*   Trying 127.0.0.1:8080...
* Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
> GET / HTTP/1.1
> Host: 127.0.0.1:12345
> User-Agent: curl/7.81.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 429 Too Many Requests
< denial-reason: ipv4!
< content-length: 0
<
* Connection #0 to host 127.0.0.1 left intact
```

**Comment by capflam** [comment]
Here there are 2 issues. The ipv4 matching is too permissive. It succeeds if the first part of the string is a valid IPv4. It is not related to the fact there is a trailing port or not. The conversion is performed by hand and stops on the first unsupported character. If 4 valid bytes were found (0-255), a success is returned. Thus `127.0.0.1ABCD` will succeed too. 

Then, I guest we need to add a converter to remove brackets from a string ip, or something to extract an ip from the host header value. Something like `ip_only`. Or indeed, `-m ip` should match IPv6 with brackets.

**Comment by Darlelet** [comment]
>Then, I guest we need to add a converter to remove brackets from a string ip, or something to extract an ip from the host header value. Something like ip_only. Or indeed, -m ip should match IPv6 with brackets.

As 2.8.0 was just released, you might be interested by the new `rfc7239_n2nn` and `rfc7239_n2np` converters (https://docs.haproxy.org/2.8/configuration.html#7.3.1-rfc7239_n2nn).
They support 7239 node identifiers, which includes IPV6 in square brackets form.

`acl isipv6 req.hdr(host),rfc7239_n2nn -m ip ::/0`
`acl isipv4 req.hdr(host),rfc7239_n2nn -m ip 0.0.0.0/0`

Should work fine on 2.8, @Puliyo would you like to give it a try?

However you should be aware that `rfc7239_n2nn` will not work with hostnames: only ipv4,ipv6,unknown and obfuscated nodenames are supported as defined by the RFC (https://datatracker.ietf.org/doc/html/rfc7239#section-6).

**Comment by Puliyo** [comment]
Tested using 2.8.0

`haproxy.cfg`
```
global
        user nobody
        user nobody
        daemon

defaults
        mode http

frontend webserver
        bind :80
        acl isipv6 req.hdr(host),rfc7239_n2nn -m ip ::/0
        acl isipv4 req.hdr(host),rfc7239_n2nn -m ip 0.0.0.0/0
        http-request deny deny_status 400 hdr Denial-Reason "ipv4!" if isipv4
        http-request deny deny_status 400 hdr Denial-Reason "ipv6!" if isipv6
        http-request return status 204
```

`Tests`
- curl -v http://127.0.0.1:80 -H 'host: 1.1.1.1' --> 400 "ipv4!"
- curl -v http://127.0.0.1:80 -H 'host: 1.1.1.1:12345' --> 400 "ipv4!"
- curl -v http://127.0.0.1:80 -H 'host: 1.1.1.1:abc' --> 204
- curl -v http://127.0.0.1:80 -H 'host: 2001:db8:85a3:8d3:1319:8a2e:370:7348' --> 204
- curl -v http://127.0.0.1:80 -H 'host: [2001:db8:85a3:8d3:1319:8a2e:370:7348]' --> 400 "ipv6!"
- curl -v http://127.0.0.1:80 -H 'host: [2001:db8:85a3:8d3:1319:8a2e:370:7348]:12345' --> 400 "ipv6!"
- curl -v http://127.0.0.1:80 -H 'host: [2001:db8:85a3:8d3:1319:8a2e:370:7348]:abc' --> 204
- curl -v http://127.0.0.1:80 -H 'host: example.com' --> 204
- curl -v http://127.0.0.1:80 -H 'host: example.com:80' --> 204

Working great.
Just that rfc7239#section-6 syntax does not capture ipv6 without bracket, so 204 returns for 'host: 2001:db8:85a3:8d3:1319:8a2e:370:7348'
This can be worked around with combination of rule with and without rfc7239_n2nn.

```
acl isipv6_1 req.hdr(host) -m ip ::/0
acl isipv6_2 req.hdr(host),rfc7239_n2nn -m ip ::/0
http-request deny deny_status 400 hdr Denial-Reason "ipv6!" if isipv6_1 || isipv6_2
...
curl -v http://127.0.0.1:80 -H 'host: 2001:db8:85a3:8d3:1319:8a2e:370:7348' --> 400 "ipv6!"
```

**Comment by Darlelet** [comment]
Great, thanks for the feedback!

>Just that rfc7239#section-6 syntax does not capture ipv6 without bracket, so 204 returns for 'host: 2001:db8:85a3:8d3:1319:8a2e:370:7348'

Shouldn't IPV6 addresses always be enclosed in square brackets under HTTP context for `host` and friends anyway?
RFC7239 doesn't allow this form but I don't know about other RFCs that might allow it.

If this is the case, then maybe we could try to improve `hdr_ip` since it seems to be the proper tool for this
kind of job, and try to support all known IPV4/IPV6 forms from multiple RFCs (with and without optional port)?

Indeed, while the `hdr_ip` documentation stipulates that:
```
The parser strictly adheres to the format
described in RFC7239, with the extension that IPv4 addresses may optionally
be followed by a colon (':') and a valid decimal port number (0 to 65535)
```

The `[ipv6]:port` form (7239 compliant) doesn't work with `hdr_ip` for now:
- curl -v [http://127.0.0.1:80](http://127.0.0.1/) -H 'host: [2001:db8:85a3:8d3:1319:8a2e:370:7348]:12345' --> 204

So the `hdr_ip` converter's description remains ambiguous because of partial 7239 support (7b0e00d943feaa6320522c376fbbff18965b1345 + d1773e6881caba229575fa7e0525846697dd4e21), and IPV6 form without bracket shouldn't be allowed if the parser "strictly adheres to the 7239 format" in my opinion.

**Comment by Puliyo** [comment]
You are right. Have not found any doc stating ipv6 in host header can be without bracket.

https://tools.ietf.org/html/rfc7230#section-5.4
https://tools.ietf.org/html/rfc3986#section-3.2.2
https://tools.ietf.org/html/rfc7239#section-6.1
RFC2373: https://stackoverflow.com/questions/7468782/http-request-over-ipv6

**Comment by capflam** [comment]
Is the problem fixed or is there anything left to do?

**Comment by Darlelet** [comment]
TLDR: 

`hdr_ip` could benefit from better rfc7239 support
Although `req.hdr(host),rfc7239_n2nn` may be used as a workaround since 2.8 for strict rfc7239 compliance

**Comment by capflam** [comment]
Copy that ! It seems reasonable to close the issue and keep a note on `hdr_ip` improvement. There is nothing much to add here. Thanks !
